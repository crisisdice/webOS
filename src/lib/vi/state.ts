
export enum VI_STATE {
  EDIT,
  INSERT
}
