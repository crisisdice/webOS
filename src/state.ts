import { standardUp, standardDown } from './mappings'
import { DASH, EMPTY } from './lib/constants'
import type { ShellState } from './lib/types'

export const EMPTY_LINE = Object.freeze({
  PRECARET: EMPTY,
  CARET: DASH,
  POSTCARET: EMPTY,
  CARET_ACTIVE: false,
})

export const INITAL_STATE: Readonly<ShellState> = Object.freeze({
  OLD_COMMANDS: [],
  COMMAND_LINE: { ...EMPTY_LINE },
  USER : 'guest',
  PWD : '/home/guest',
  CONTROL_DOWN: false,
  FULL_SCREEN: false,
  NAME: 'sh',
  UP_MAPPING : standardUp,
  DOWN_MAPPING : standardDown,
})
