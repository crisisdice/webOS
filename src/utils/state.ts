import type { ShellState } from '../types'
import { standardUp, standardDown } from '../mappings'
import { DASH, EMPTY, PWD, HOME } from './constants'
import { setEnv, setFs } from '../lib'

export const EMPTY_LINE = Object.freeze({
  PRECARET: EMPTY,
  CARET: DASH,
  POSTCARET: EMPTY,
  CARET_ACTIVE: false,
})

export const INITAL_STATE: Readonly<ShellState> = Object.freeze({
  OLD_COMMANDS: [],
  COMMAND_LINE: { ...EMPTY_LINE },
  USER: 'guest',
  PWD: '/home/guest',
  CONTROL_DOWN: false,
  FULL_SCREEN: false,
  NAME: 'sh',
  UP_MAPPING: standardUp,
  DOWN_MAPPING: standardDown,
})

// TODO let init handle user and pwd
// TODO whoami and login
export const init = () => {
  setEnv(PWD, '/home/guest')
  setEnv(HOME, '/home/guest')

  setFs({
    home: {
      guest: {
        '.bashrc': '#!/usr/sh',
        '.history': '',
      },
      root: {},
    },
  })
  console.log('initalized')
}
