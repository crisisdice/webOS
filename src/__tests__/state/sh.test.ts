import { runAll } from '../state'
import { sh } from '../cases'

runAll(sh)
