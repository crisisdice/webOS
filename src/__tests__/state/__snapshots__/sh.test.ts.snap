// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LEFT 3><RIGHT 3> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`<LEFT 3><RIGHT 3> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`<LEFT 3><RIGHT 3> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`echo<ENTER>ls<ENTER>ArrowUpArrowUpArrowDown 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "ls",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": 0,
  "NAME": 0,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "echo",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "ls",
      },
      "PWD": "/home/guest",
      "STDOUT": ".bashrc .history",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`echo<ENTER>ls<ENTER>ArrowUpArrowUpArrowDown 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`echo<ENTER>ls<ENTER>ArrowUpArrowUpArrowDown 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`g<LEFT 1> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "g",
    "CARET_WIDTH": 1,
    "EOL": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`g<LEFT 1> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`g<LEFT 1> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 1><RIGHT 2> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "hello",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 1><RIGHT 2> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 1><RIGHT 2> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 2><RIGHT 1> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "o",
    "CARET_WIDTH": 1,
    "EOL": false,
    "POSTCARET": "",
    "PRECARET": "hell",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 2><RIGHT 1> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 2><RIGHT 1> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 3><RIGHT 3> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "hello",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 3><RIGHT 3> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 3><RIGHT 3> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 5><RIGHT 6> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "hello",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 5><RIGHT 6> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 5><RIGHT 6> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 5><RIGHT 7> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "hello",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 5><RIGHT 7> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 5><RIGHT 7> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 6><RIGHT 4> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "o",
    "CARET_WIDTH": 1,
    "EOL": false,
    "POSTCARET": "",
    "PRECARET": "hell",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 6><RIGHT 4> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 6><RIGHT 4> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 7><RIGHT 7> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "hello",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 7><RIGHT 7> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 7><RIGHT 7> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;
